dds/idl/marshal_generator.cpp

```diff
@@ -769,567 +1108,253 @@ namespace {
           "    return false;\n"
           "  }\n";
       }
-      be_global->impl_ << "  return true;\n";
+      be_global->impl_ <<
+        "  return true;\n";
+    }
+  }
+
+  void gen_sequence(UTL_ScopedName* name, AST_Sequence* seq)
+  {
+    gen_sequence_i(name, seq, false);
+    if (needs_nested_key_only(seq)) {
+      gen_sequence_i(name, seq, true);
     }
   }
 
   void gen_anonymous_sequence(const FieldInfo& sf)
   {
-    string cxx = sf.name_;
-    if (!sf.as_act_->in_main_file()) {
-      if (sf.as_act_->node_type() == AST_Decl::NT_pre_defined) {
-        if (be_global->language_mapping() != BE_GlobalData::LANGMAP_FACE_CXX &&
-            be_global->language_mapping() != BE_GlobalData::LANGMAP_SP_CXX) {
-          be_global->add_include(("dds/CorbaSeq/" + nameOfSeqHeader(sf.as_act_)
-                                  + "SeqTypeSupportImpl.h").c_str(), BE_GlobalData::STREAM_CPP);
-        }
-      } else {
-        be_global->add_referenced(sf.as_act_->file_name().c_str());
-      }
+    gen_sequence_i(0, 0, false, &sf);
+    if (needs_nested_key_only(sf.type_)) {
+      gen_sequence_i(0, 0, true, &sf);
+    }
+  }
+
+  void gen_array_i(
+    UTL_ScopedName* name, AST_Array* arr, bool nested_key_only, const FieldInfo* anonymous = 0)
+  {
+    if (anonymous) {
+      arr = dynamic_cast<AST_Array*>(anonymous->type_);
     }
+    const std::string named_as = anonymous ? anonymous->scoped_type_ : scoped(name);
+    Wrapper base_wrapper(arr, named_as, "arr");
+    base_wrapper.nested_key_only_ = nested_key_only;
+    if (!anonymous) {
+      be_global->add_include("dds/DCPS/Serializer.h");
+    }
+    NamespaceGuard ng(!anonymous);
 
     const bool use_cxx11 = be_global->language_mapping() == BE_GlobalData::LANGMAP_CXX11;
-    const bool primitive = sf.as_cls_ & CL_PRIMITIVE;
-    const string check_empty = use_cxx11 ? "seq.empty()" : "seq.length() == 0";
-    const string get_length = use_cxx11 ? "static_cast<uint32_t>(seq.size())" : "seq.length()";
-    const string get_buffer = use_cxx11 ? "wrap.val_->data()" : "seq.get_buffer()";
-    string const_cxx = cxx;
-    if (use_cxx11) {
-      be_global->header_ << "struct " << sf.underscored_ << "_tag {};\n\n";
-      const_cxx = "IDL::DistinctType<const " + cxx + ", " + sf.underscored_ + "_tag>";
-    } else {
-      const_cxx = "const " + cxx + '&';
+
+    AST_Type* elem = resolveActualType(arr->base_type());
+    TryConstructFailAction try_construct = be_global->array_element_try_construct(arr);
+    Classification elem_cls = classify(elem);
+    const bool primitive = elem_cls & CL_PRIMITIVE;
+    if (!elem->in_main_file()
+        && elem->node_type() != AST_Decl::NT_pre_defined) {
+      be_global->add_referenced(elem->file_name().c_str());
+    }
+    const std::string cxx_elem =
+      anonymous ? anonymous->scoped_elem_ : scoped(arr->base_type()->name());
+    size_t n_elems = 1;
+    for (size_t i = 0; i < arr->n_dims(); ++i) {
+      n_elems *= arr->dims()[i]->ev()->u.ulval;
```

Comment:
This appears in at least 4 other places, we should have a common function for it.